### ISSUE TITLE
Specify the format and behavior of the relatedSpdxElement for JSON, YAML and XML formats

### ISSUE THREAD

**Issue by goneall** [spdx/spdx-spec#96]
Section 7.1 specifies the format for relationships for tag/value and RDF.

The tag/value is a parseable string of the format ` [“DocumentRef-”[idstring]”:”]SPDXID <relationship> [“DocumentRef-”[idstring]”:”]SPDXID` and the RDF specifies a related SPDX element as a URI reference to another SPDX element.

Neither of these formats work well for the JSON, YAML and XML formats proposed.

For JSON, YAML, and XML it would be preferable to have a property/element to describe the related element.  SInce relationships can represent an undirected graph with possible cycles, we can not, however, refer to the actual object without creating problems for the tools (or infinitely sized documents).

**Comment by goneall** [comment]
Propose that we define a property in the Relationship object/class/type `relatedSpdxElement` with a definition borrowed from section 5.2 snippet from file SPDX identifier:

```
Cardinality: Mandatory, one.
DataFormat: [“DocumentRef-”[idstring]”:”] SPDXID
where DocumentRef-[idstring]: is an optional reference to an external SPDX document as described in section 2.6
where SPDXID is a string containing letters, numbers, “.”,“-”. asdescribed in sections (2.3, 3.2, 4.2).
```

By keeping the property name the same as RDF, we make it easier to translate between RDF and these other formats even though the definition of the field is slightly different.

By use the SPDX ID, we can reference related SPDX elements both inside the SPDX document and externally.  There is also a well defined mapping to the actual objects we can use to implement the actual references.

**Comment by mcjaeger** [comment]
The SPDX file for the gcc which is hilariously large I think another representation for:
````
file path, file name, file hash, license, copyright
````
would be useful for tooling. I think, still it is a matter of technical constraints, because parsing 100MB of RDF lets raise the memory consumption easily by factor 10 on Java (sw360) and PHP (fossology) at least. Note that we did not consider an XML streaming parser because of RDF. That is not nice for a server application.
So a reference for another file format would be very good, however, I am not sure the RDF graph is useful as data structure for the plain file-license information.

**Comment by swinslow** [comment]
Hi @goneall, I might be thinking about this entirely the wrong way  =)  But for something like the JSON representation, would this be made easier by pulling Relationships and Annotations "out of" the Package and File sections, and just having a single "relationships" array that is global to the document?

Not sure I phrased that well, so here's an example. When I look at the [sample JSON file](https://github.com/spdx/spdx-spec/blob/fd555f596c0b31cc13de5834b1f1a53614474132/examples/SPDXJSONExample-v2.0.json), at [line 113](https://github.com/spdx/spdx-spec/blob/fd555f596c0b31cc13de5834b1f1a53614474132/examples/SPDXJSONExample-v2.0.json#L113) I see the following:

```
      "Package" : {
        "name" : "glibc",
        "id" : "SPDXRef-Package",
        . . .
        "relationships" : [ {
          "relationshipType" : "DYNAMIC_LINK",
          "comment" : null,
          "relatedSpdxElementId" : "SPDXRef-Saxon"
        }, {
          "relationshipType" : "CONTAINS",
          "comment" : null,
          "relatedSpdxElementId" : "SPDXRef-JenaLib"
        } ],
        . . .
```

I guess I'm wondering, should those relationships actually be an attribute within the `Package` data structure? Each Relationship in an SPDX document has two elements that it relates to. If both elements are described in the same SPDX document, then it isn't clear to me that that Relationship should necessarily be "owned" by the first element. 

And, taking it a step further, I think it could be valid to express relationships where _both_ elements are external -- where neither element is necessarily described in this document. (Perhaps silly, but I could imagine a non-frivolous use case for this.)

So I guess I'm wondering, instead of the above, what if the JSON document had a single, "global" `relationships` array at the end of the document, that looked something like this:

```
        "relationships" : [ {
          "firstElementId": "SPDXRef-Package"
          "relationshipType" : "DYNAMIC_LINK",
          "comment" : null,
          "secondElementId" : "SPDXRef-Saxon"
        }, {
          "firstElementId": "SPDXRef-Package"
          "relationshipType" : "CONTAINS",
          "comment" : null,
          "secondElementId" : "SPDXRef-JenaLib"
        },
        . . .
],
```

and then the same sort of approach for the other formats.

**Comment by goneall** [comment]
> I guess I'm wondering, should those relationships actually be an attribute within the Package data structure? Each Relationship in an SPDX document has two elements that it relates to. If both elements are described in the same SPDX document, then it isn't clear to me that that Relationship should necessarily be "owned" by the first element.

In the [SPDX object model](https://wiki.spdx.org/view/Technical_Team/Model_2_2), the relationship is a property of the SPDX element.  The JSON representation should follow the model.

Similary, Annotations are properties of the SPDX Element.

@swinslow you raise some very good points, that they are probably issues with the object model rather than the JSON implementation.  There are other issues with the current model which would be solved by your proposal such as infinite recursion when reconstituting the objects from the relationship properties.  There is a bunch of ugly code in the Java tools to work around this issue.

I would support changing the object model, but it would be a breaking change and something we should consider for 3.0 rather than 2.2.

**Comment by swinslow** [comment]
Hi @goneall, thanks and good point. I hadn't looked at this in terms of the SPDX data / object model in the appendix.

I was about to write up a lengthier response, with questions about how this interacts with the Data Format for Relationships in 7.1.4, but in drafting it I realized (1) I don't know much about RDF and (2) it might be easier to discuss live briefly on the next tech call. I'll plan to bring it up then if that's okay.  =)  Thanks!

**Comment by zvr** [comment]
The whole point of XML/RDF is that data are expressed in _triplets_ of the form: subject - predicate - object. See https://en.wikipedia.org/wiki/Semantic_triple.
Therefore you have a Package (or another item, like a file), which has a relationship (e.g., contains) to another object (again, a package or a file or ...).
Going to a top-level relationships array destroys all this structural convention.

Actually, @swinslow , to your question why this "relationship" is owned by the first element, the answer is: it shouldn't! Typically in RDF data, if you have the triplet A-contains-B ("owned by A", in your parlance), you also have B-contained_in-A in the other object.
I assume you also have to plan all these relationships in the toplevel array; otherwise one would need external info which are reciprocal to search for both first and second elements, etc.

**Comment by goneall** [comment]
Consensus on the 15 Oct 15 call:

Move relationships to a separate array within the document with the format:

```
        "relationships" : [ {
          "subjectElementId": "SPDXRef-Package"
          "relationshipType" : "DYNAMIC_LINK",
          "comment" : null,
          "objectElementId" : "SPDXRef-Saxon"
        }, {
          "subjectElementId": "SPDXRef-Package"
          "relationshipType" : "CONTAINS",
          "comment" : null,
          "objectElementId" : "SPDXRef-JenaLib"
        },
        . . .
```

Note - the use of terms subject and object are consistent with the RDF definitions to describe the semantics of these terms.
],

**Comment by goneall** [comment]
I would like to propose a slight change in the above proposal.

Instead of `objectElementId`, use `relatedSpdxElementId`.  Reason: `relatedSpdxElementId` is already used in the current tag/value and RDF formats.  Using `relatedSpdxElementId` would make the JSON and  YAML formats easier to convert between RDF, YAML, JSON and tag/value.

@zvr , @swinslow let me know if you disagree.

**Comment by swinslow** [comment]
@goneall that makes sense to me!

**Comment by zvr** [comment]
Do you also want the subject to be `spdxElementId` ?
No need to keep "subject" when we have X - relationship - related.

**Comment by goneall** [comment]
> Do you also want the subject to be spdxElementId ?

I agree `spdxElementId` is a better term.

I updated the PR with the new term.

**Comment by goneall** [comment]
Resolved with PR #197
